### PR TITLE
Rule registry

### PR DESCRIPTION
**Pull request by lixitrixi**
Here I adapt the rule registry system I developed at https://github.com/lixitrixi/antwort, which was itself developed with input from @niklasdewally.

Currently, using the `register_rule` macro requires the presence of the `conjure_core` crate and `conjure_rules::_RULES_DISTRIBUTED_SLICE` in the current scope. One possible future project could be to clean up the import structure and macro definition to allow using it right out of the box with one import from `conjure_oxide`. For now, I believe this system will suffice as we create the rewriting system.

### CODE REVIEW

**Comment by lixitrixi** [comment]
`cargo test` is currently failing for me on `main`, I will investigate and hopefully add a fix here

**Review by niklasdewally — COMMENTED** [review]
I don't have much to add that isn't already in our DM's, but for completeness and Oz/Chris, here are my thoughts:

* Syntax and functionality all good, and adding rules seems nice and intuitive.

*  I don't quite get the split of things between crates anymore. Which crates are needed for compilation / circular dependency reasons, and which are "aesthetic"? 

     (for context, crates with proc-macros can only publicly export proc-macros. There are also issues to do with macro expansion and using things from external crates inside the macro, and these are made worse by this limitation) 

    This should also maybe be clearly documented in READMEs for each internal crate -  especially in cases where a macro crate just exists for proc-macro compilation reasons, and should never be used directly, even internally inside this project.
    
     If there is no more circular dependency in the proc-macro stuff, maybe we don't need conjure-core anymore? My understanding is there needs to be foo and a foo-proc crate where foo reexports foo-proc::macro as foo::macro. In this scenario the macro expands in the context of foo, so can access stuff defined in foo without any problems.

* I understand that the linkme hack is required for reasons outside of your control, but what are the implications of this? Do all rules need to be in the same crate? Same file? We could maybe look into alternatives to linkme too.

* Can a user add more rules in an external crate? Linkme not withstanding, can the macro find the imports it needs when used inside another crate?

**Comment by niklasdewally** [comment]
(Note that this is stacked on top of #137, so that should be merged first, and the first two commits here are from there.)

**Comment by lixitrixi** [comment]
> - I understand that the linkme hack is required for reasons outside of your control, but what are the implications of this? Do all rules need to be in the same crate? Same file? We could maybe look into alternatives to linkme too.

As long as you have `conjure_rules::_RULES_DISTRIBUTED_SLICE` in scope, the macro will register the function as a rule.

**Comment by niklasdewally** [comment]
> > - I understand that the linkme hack is required for reasons outside of your control, but what are the implications of this? Do all rules need to be in the same crate? Same file? We could maybe look into alternatives to linkme too.
> 
> As long as you have `conjure_rules::_RULES_DISTRIBUTED_SLICE` in scope, the macro will register the function as a rule.

I was referring to this:

```rs
static _TEMP: &Rule = &rules::CONJURE_GEN_RULE_IDENTITY; // Temporary hack to force the static to be included in the binary
```

I remember it working out to something like that every file with rules in has to be mentioned in lib rs for rust.to actually bother.to compile it? Is that correct?

**Comment by lixitrixi** [comment]
> I can't remember the details, but isnt this hack required on Mac so that every file with rules in has to be mentioned in lib rs for rust.to actually bother.to compile it? Is that correct?

Yes, that's what's going on

**Comment by niklasdewally** [comment]
> > I can't remember the details, but isnt this hack required on Mac so that every file with rules in has to be mentioned in lib rs for rust.to actually bother.to compile it? Is that correct?
> 
> Yes, that's what's going on

So every file containing rules needs to be in lib.rs to be recognised?

**Comment by niklasdewally** [comment]
To clarify, this is where that hack comes from, and why it's needed:

https://github.com/dtolnay/linkme/issues/31

**Comment by niklasdewally** [comment]
> > > I can't remember the details, but isnt this hack required on Mac so that every file with rules in has to be mentioned in lib rs for rust.to actually bother.to compile it? Is that correct?
> > 
> > 
> > Yes, that's what's going on
> 
> So every file containing rules needs to be in lib.rs to be recognised?


I think i've got a patch that removes this hack, and allows rules to be defined anywhere without having to `use` the module in question.

Sent it to Felix to test on his machine, and he will push it if it works :)

**Comment by niklasdewally** [comment]
> One possible future project could be to clean up the import structure and macro definition to allow using it right out of the box with one import from conjure_oxide

I'm currently discussing this with Felix - not ready to merge yet :)

**Review by niklasdewally — COMMENTED** [review]
After more discussion with Felix, and some patches, I think the major issues discussed previously are resolved.

Rules can now be defined anywhere in any crate without any issues, or extra "quirks" implementers needs to know about. This also applies to external crates!!

This was not possible before to two issues (corresponding to my 2 commits):
* a `linkme` compilation bug.
* macros being unhygenic, and just token substitution, and this affecting how imports and dependencies work. 

(The latter was a rather interesting problem, see my commit message [here](https://github.com/conjure-cp/conjure-oxide/pull/141/commits/d30c2be3475945c3350c83dbf629ebe53da9581a) for the hows and whys of the final project layout.)


----

I think there could be one more commit with doc comments for `conjure_rules`, then this would be done!

**Comment by lixitrixi** [comment]
Do we think that doctests are worthwhile or even possible in this context? I'm not sure how Cargo testing will interact with linker shenanigans, and I am running into constant import issues while attempting to fix this problem.

@niklasdewally

**Comment by niklasdewally** [comment]
> Do we think that doctests are worthwhile or even possible in this context? I'm not sure how Cargo testing will interact with linker shenanigans, and I am running into constant import issues while attempting to fix this problem.
> 
> @niklasdewally 

I think one should always write an example if possible.

Doctests "pass" if they compile and run without panicking - the important thing I think is making the examples compile so rust will tell us when they no longer compile and we need to update them. Asserts arnt too important I would say.

File: crates/conjure_rules/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
+//! ### A decentralised rule registry for Conjure Oxide
+//!
+//! This crate allows registering valid functions as expression-reduction rules.
+//! Functions can be decorated with the `register_rule` macro in any downstream crate and be used by Conjure Oxide's rule engine.
+//! To achieve compile-time linking, we make use of the [`linkme`](https://docs.rs/linkme/latest/linkme/) crate.
+//!
```
Comment (niklasdewally) [inline crates/conjure_rules/src/lib.rs:6]: (Pedantic): This is implementation detail so should not be in the user-facing doc strings.

File: crates/conjure_rules/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    pub use linkme::distributed_slice;
+}
+
+#[doc(hidden)]
+#[distributed_slice]
+pub static RULES_DISTRIBUTED_SLICE: [Rule<'static>];
+
+/// Returns a copied `Vec` of all rules registered with the `register_rule` macro.
+///
+/// Rules defined in the same file will remain contiguous and in order, but order may not be maintained between files.
```
Comment (niklasdewally) [inline crates/conjure_rules/src/lib.rs:40]: I don't think this is something we should let consumers of the rule_registry rely on, and have to hold ourselves to - this is heavily dependent on the fact that we use linkme in the implementation, which itself depends heavily on rust version and underlying compiler tech and LLVM stuff beyond our understanding or control

File: crates/conjure_rules_proc_macro/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Ident, ItemFn};
+
+#[doc(hidden)]
+#[proc_macro_attribute]
+/// This procedural macro registers a decorated function with `conjure_rules`' global registry.
+/// It may be used in any downstream crate. For more information on linker magic, see the [`linkme`](https://docs.rs/linkme/latest/linkme/) crate.
+///
```
Comment (niklasdewally) [inline crates/conjure_rules_proc_macro/src/lib.rs:9]: (Pedantic) we should treat this crate as internal to `conjure_rules`, so no-one else should use / see this but us

**Review by niklasdewally — APPROVED** [review]
LGTM!

(just some subjective documentation stuff)

File: crates/conjure_rules_proc_macro/src/lib.rs (hunk 3 above)
Comment (lixitrixi) [reply crates/conjure_rules_proc_macro/src/lib.rs:9]: The macro itself will be externally visible if `conjure_rules` is imported, no?

Comment (niklasdewally) [reply crates/conjure_rules_proc_macro/src/lib.rs:9]: through `conjure_rules` only :)

Comment (niklasdewally) [reply crates/conjure_rules_proc_macro/src/lib.rs:9]: unless the user explicitly imports `conjure_rules_proc_macro`, which we should tell them not to do in the description for conjure_rules_proc-macro.

Comment (niklasdewally) [reply crates/conjure_rules_proc_macro/src/lib.rs:9]: btw, you can directly document the re-export in `conjure_rules`, you dont need to put the docs here.

**Comment by lixitrixi** [comment]
@ozgurakgun This should now be good to merge (after #137)! I've started the manual rewrite proof of concept and will have something to show tomorrow.

**Pull request merged**